build-system/tasks/helpers.js

```diff
@@ -374,96 +371,119 @@ function finishBundle(srcFilename, destDir, destFilename, options) {
-      const bundleComplete = performBundle(/* continueOnError */ true);
+    const watchFunc = async () => {
+      const bundleDone = await performBundle(/* continueOnError */ true);
       if (options.onWatchBuild) {
-        options.onWatchBuild(bundleComplete);
+        options.onWatchBuild(bundleDone);
       }
     };
-    bundler = watchify(bundler);
-    bundler.on('update', debounce(watchFunc, watchDebounceDelay));
+    watch(entryPoint).on('change', debounce(watchFunc, watchDebounceDelay));
   }
 
   /**
-   * @param {boolean} continueOnError
-   * @return {Promise}
+   * Babel plugin for esbuild
    */
-  function performBundle(continueOnError) {
-    let startTime;
-    return toPromise(
-      bundler
-        .bundle()
-        .once('readable', () => (startTime = Date.now()))
-        .on('error', (err) =>
-          handleBundleError(err, continueOnError, destFilename)
-        )
-        .pipe(source(srcFilename))
-        .pipe(buffer())
-        .pipe(sourcemaps.init({loadMaps: true}))
-        .pipe(
-          regexpSourcemaps(
-            /\$internalRuntimeVersion\$/g,
-            internalRuntimeVersion,
-            'runtime-version'
-          )
-        )
-        .pipe(regexpSourcemaps(/([^]+)/, devWrapper, 'wrapper'))
-        .pipe(rename(destFilename))
-        .pipe(sourcemaps.write('.'))
-        .pipe(gulp.dest(destDir))
-        .on('end', () =>
-          finishBundle(srcFilename, destDir, destFilename, options)
-        )
-    )
-      .then(() => {
-        let name = destFilename;
-        if (options.latestName) {
-          const latestMaxName = options.latestName.split('.js')[0] + '.max.js';
-          name = `${name} → ${latestMaxName}`;
-        }
-        endBuildStep('Compiled', name, startTime);
-      })
-      .then(() => {
-        const target = path.basename(destFilename, path.extname(destFilename));
-        if (UNMINIFIED_TARGETS.includes(target)) {
-          return applyAmpConfig(
-            `${destDir}/${destFilename}`,
-            /* localDev */ true,
-            /* fortesting */ options.fortesting
-          );
+  const esbuildBabelPlugin = {
+    name: 'babel',
+    async setup(build, {transform} = {}) {
+      const updateVersion = (contents) => {
+        return contents.replace(
+          /\$internalRuntimeVersion\$/g,
+          internalRuntimeVersion
+        );
+      };
+
+      const transformContents = async ({file, contents}) => {
+        const babelOptions = babel.loadOptions({
+          caller: {name: 'unminified'},
+          filename: file.path,
+          sourceFileName: path.relative(process.cwd(), file.path),
+        });
+        const result = await babel.transformAsync(contents, babelOptions);
+        return {contents: updateVersion(result.code)};
+      };
+
+      if (transform) {
+        return await transformContents(transform);
+      }
+
+      build.onLoad({filter: /.*/, namespace: ''}, async (file) => {
+        if (cache.has(file.path)) {
```

Comment:
how are we cache busting?

Reply:
Oooh, excellent point. We weren't busting the cache earlier, but I've added code that does so and tested it as follows:

```sh
# Clean build state
gulp clean

# Launch lazy-build dev server
gulp

# Load http://127.0.0.1:8000/examples/article.amp.html
# Prints `Powered by AMP ⚡ HTML – Version `

# Modify `src/amp.js` and change the line to `Powered by AMP2`
# `amp.js` is rebuilt

# Reoad http://127.0.0.1:8000/examples/article.amp.html
# Prints `Powered by AMP2 ⚡ HTML – Version `
```

Incidentally, this is the exact same mechanism used by `pre-closure-babel.js` used by `gulp dist`. I'll revisit the caching there when I refactor `gulp dist` in case we need to do the same cache busting there too.

Reply:
Sounds great. Just tested it locally and it worked!